plugins/modem-manager/fu-mbim-qdu-updater.c

```diff
@@ -0,0 +1,464 @@
+/*
+ * Copyright (C) 2021 Jarvis Jiang <jarvis.w.jiang@gmail.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <string.h>
+#include <sys/stat.h>
+#include <stdio.h>
+
+#include "fu-common.h"
+#include "fu-mbim-qdu-updater.h"
+#include "fu-chunk.h"
+#include "fu-mm-utils.h"
+
+#if MBIM_CHECK_VERSION(1,25,3)
+#define FU_MBIM_QDU_MAX_OPEN_ATTEMPTS 8
+
+struct _FuMbimQduUpdater {
+	GObject		 parent_instance;
+	gchar		*mbim_port;
+	MbimDevice	*mbim_device;
+};
+
+G_DEFINE_TYPE (FuMbimQduUpdater, fu_mbim_qdu_updater, G_TYPE_OBJECT)
+
+typedef struct {
+	GMainLoop	*mainloop;
+	MbimDevice	*mbim_device;
+	GError		*error;
+	guint		 open_attempts;
+} OpenContext;
+
+static void fu_mbim_qdu_updater_mbim_device_open_attempt (OpenContext *ctx);
+
+static void
+fu_mbim_qdu_updater_mbim_device_open_abort_ready (GObject *mbim_device, GAsyncResult *res, gpointer user_data)
+{
+	OpenContext *ctx = (OpenContext *) user_data;
+
+	g_warn_if_fail (ctx->error != NULL);
+
+	/* ignore errors when aborting open */
+	mbim_device_close_finish (MBIM_DEVICE (mbim_device), res, NULL);
+
+	ctx->open_attempts--;
+	if (ctx->open_attempts == 0) {
+		g_clear_object (&ctx->mbim_device);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	/* retry */
+	g_clear_error (&ctx->error);
+	fu_mbim_qdu_updater_mbim_device_open_attempt (ctx);
+}
+
+static void
+fu_mbim_qdu_updater_open_abort (OpenContext *ctx)
+{
+	mbim_device_close (ctx->mbim_device,
+				10, NULL, fu_mbim_qdu_updater_mbim_device_open_abort_ready, ctx);
+}
+
+static void
+fu_mbim_qdu_updater_mbim_device_open_ready (GObject *mbim_device, GAsyncResult *res, gpointer user_data)
+{
+	OpenContext *ctx = (OpenContext *) user_data;
+
+	if (!mbim_device_open_finish (MBIM_DEVICE (mbim_device), res, &ctx->error)) {
+		fu_mbim_qdu_updater_open_abort (ctx);
```

Comment:
A failed open doesn't require to explicitly close. This logic also seems copied from the FuQmiPdcUpdater logic, but in that case it made sense, as we may have failed during the QMI client allocation, so that would have required an explicit QMI device close. In this case, it's not needed, you can fully skip the need to `mbim_device_close()` if the `open_finish()` returns failure.

Reply:
But you would need to move the "open_attempts" logic to retry here though, please don't lose that

Reply:
Not really, you still need to see if `mbim_device_open_finish()` failed. The open attempts logic should be used only if `mbim_device_open_finish() fails`.

Reply:
Not really, you still need to see if `mbim_device_open_finish()` failed. The open attempts logic should be used only if `mbim_device_open_finish() fails`.

Reply:
OK, thanks for your reply.